### PR TITLE
Add fmha_v2 attention backend for SM90/120

### PR DESCRIPTION
**Pull request by akhilg-nv**
## Motivation

Enables TRT-LLM's fmha_v2 prefill attention backend for SM90/120. This backend is more performant than the current default (FA3), and also enables skip-softmax feature, which can further boost performance for long-context / applications where attention output is sparse.

## Modifications

Baseline enablement of trtllm's fmha_v2 prefill kernel for SM90/120. https://github.com/sgl-project/sglang/pull/32272 is needed for further perf improvement, and https://github.com/sgl-project/sglang/pull/32268 / https://github.com/sgl-project/sglang/pull/32269 are used for the MTP accuracy reporting.

## Accuracy Tests
```
python3 -m sglang.launch_server \
  --model /SGL/models/Qwen3-Next-80B-A3B-Thinking-FP8 \
  --tp 4 --mem-fraction-static 0.8 \
  --disable-radix-cache --mamba-ssm-dtype float32 \
  --moe-runner-backend deep_gemm --attention-backend trtllm_mha \
  --trust-remote-code --port 30000
# --kv-cache-dtype fp8_e4m3
```

| Config | KV dtype | Score | Repeat range |
|---|---|---|---|
| FA3 | bf16 | 0.725 | 0.712-0.742 | 
| fmha_v2 (trtllm_mha) | bf16 | 0.723 | 0.702-0.753 | 
| FA3 + MTP | bf16 | 0.734 | 0.712-0.758 | 
| fmha_v2 prefill + MTP (fa3 verify) | bf16 | 0.721 | 0.707-0.732 | 
| FA3 | **fp8_e4m3** | 0.727 | 0.707-0.747 | 
| fmha_v2 (trtllm_mha) | **fp8_e4m3** | 0.724 | 0.682-0.753 | 

### MTP

Add to server launch: `--speculative-algorithm NEXTN --speculative-num-steps 3 \
--speculative-eagle-topk 1 --speculative-num-draft-tokens 4`

| Config | Accept len | Accept rate | GPQA-20 score |
|---|---|---|---|
| FA3 + MTP (bf16) | 2.37 (2.35) | 0.46 (0.45) | 0.750 |
| fmha_v2 prefill + MTP (bf16) | 2.45 (2.30) | 0.48 (0.43) | 0.700 |
| pure trtllm_mha + MTP w/ XQA mask fix (bf16, 10q) | 2.20 | 0.40 | 0.70 |

## Speed Tests and Profiling

The e2e performance is not improved with just this PR because it requires a large copy operation before the compute kernel to interleave KV. https://github.com/flashinfer-ai/flashinfer/pull/3575 and https://github.com/sgl-project/sglang/pull/32272 address this issue.

Unit Kernel Measurements show ~15% improvement over FA3 over a variety of problem sizes, reaching 20% at small problems.

image

### Skip-Softmax

This feature can provide additional speedup, especially for long-context inference where the attention logits are sparse. We define the threshold factor `\lambda := a / L` where a is a hyperparameter, and L is the sequence length. To demonstrate the potential speedup, we keep a = 1000 here for simplicity. 

| ISL  | Effective threshold (1000 / ISL) | bf16 gain | fp8 gain | 
| --- | --- | --- | --- |
| 8k | 0.122 | +1.8% | +0.2% |
| 32k | 0.031 | +5.2% | +1.7% |
| 128k | 0.008 | +12.9% | + 3.5% | 
          

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #32205487242](https://github.com/sgl-project/sglang/actions/runs/32205487242)
Latest PR Test (Extra): :x: [Run #32205486818](https://github.com/sgl-project/sglang/actions/runs/32205486818)

### CODE REVIEW

**Comment by nvpohanh** [comment]
@akhilg-nv could you fix the merge conflict? thanks!

**Comment by nvpohanh** [comment]
cc @b8zhong for vis

**Comment by nvpohanh** [comment]
@akhilg-nv could you fix the conflicts? thanks

**Review by b8zhong — APPROVED** [review]

**Comment by b8zhong** [comment]
@akhilg-nv I agree with Po-Han here: https://github.com/sgl-project/sglang/pull/32272#issuecomment-5213981309

By the way, if the additional diff is not super complex and is related, we can keep always keep it it in 1 PR

**Comment by akhilg-nv** [comment]
> @akhilg-nv I agree with Po-Han here: [#32272 (comment)](https://github.com/sgl-project/sglang/pull/32272#issuecomment-5213981309)
> 
> By the way, if the additional diff is not super complex and is related, we can keep always keep it it in 1 PR

I've left it separated for now because #32272 is dependent on a flashinfer API change (https://github.com/flashinfer-ai/flashinfer/pull/4124) that has not yet merged, but this PR is ready to merge at anytime.

**Review by YAMY1234 — COMMENTED** [review]
Two SM120 compatibility notes inline.

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 1]
```diff
@@ -1342,6 +1360,30 @@ def forward_extend(
                     q_len_per_req=self.forward_metadata.max_seq_len_q,
                     multi_ctas_kv_counter_buffer=self._multi_ctas_kv_counter_buffer,
                 )
+        elif self.use_fmha_v2 and not cp_v2_active:
+            # CP-v2 must go through cp_strategy.run_attention (per-shard
+            # masking); the plain-causal fmha_v2 call below would be wrong.
+            paged_kv = torch.stack([k_cache, v_cache], dim=1)
+            o = flashinfer.prefill.trtllm_fmha_v2_prefill(
```
Comment (YAMY1234) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py:1367]: Could we add a guard or fallback for SM120 here? The pinned FlashInfer 0.6.17 explicitly rejects E4M3 queries in FMHAv2 on SM120 ([source](https://github.com/flashinfer-ai/flashinfer/blob/a0a6b019b9b27d49d209f85d028a1ae5a9b347d7/flashinfer/prefill.py#L5379-L5387)), so `--kv-cache-dtype fp8_e4m3` would fail on the first prefill.

File: python/sglang/srt/layers/attention/trtllm_mha_backend.py [hunk 2]
```diff
@@ -1342,6 +1360,30 @@ def forward_extend(
...
+                bmm2_scale=bmm2_scale,
+                batch_size=forward_batch.batch_size,
+                cum_seq_lens_q=self.forward_metadata.cu_seqlens_q,
+                cum_seq_lens_kv=self.forward_metadata.cu_seqlens_k,
+                block_tables=page_table,
+                out_dtype=self.q_data_type,
+                mask_mode="causal",
+                window_left=layer.sliding_window_size,
+                sinks=attention_sink,
+                skip_softmax_threshold_scale_factor=envs.SGLANG_SKIP_SOFTMAX_PREFILL_THRESHOLD_SCALE_FACTOR.get()
```
Comment (YAMY1234) [inline python/sglang/srt/layers/attention/trtllm_mha_backend.py:1384]: Could we disable skip-softmax on SM120 for now? In the pinned FlashInfer version, `enable_skip_softmax=True` kernels are generated only for SM90 ([source](https://github.com/flashinfer-ai/flashinfer/blob/a0a6b019b9b27d49d209f85d028a1ae5a9b347d7/flashinfer/jit/attention/fmha_v2/generator_utils.py#L6767-L6785)), so a positive threshold on SM120 can hit the unsupported-config assertion.

**Review by YAMY1234 — APPROVED** [review]
